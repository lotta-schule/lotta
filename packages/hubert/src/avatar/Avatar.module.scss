.root {
  border: 1px solid rgb(var(--lotta-accent-grey-color));
  background-color: rgb(var(--lotta-accent-grey-color));
  border-radius: 50%;
  padding: 0.5em; /* TODO: Property */
  display: inline-block;
  background-size: cover;
  box-shadow: 1px 1px 7px #0000002e; /* TODO: Property */
  height: 2em; /* TODO: Property */
  width: 2em; /* TODO: Property */
}
