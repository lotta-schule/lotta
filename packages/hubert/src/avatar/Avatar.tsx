import * as React from 'react';
import clsx from 'clsx';

import styles from './Avatar.module.scss';

export interface AvatarProps {
  /**
   * The avatar image source
   */
  src: string;

  /**
   * The title of the avatar
   */
  title?: string;

  className?: string;

  style?: React.CSSProperties;
}

export const Avatar = React.memo(
  React.forwardRef(
    (
      { src, title, className, style }: AvatarProps,
      ref: React.ForwardedRef<HTMLDivElement>
    ) => {
      return (
        <div
          role={'img'}
          title={title}
          ref={ref}
          className={clsx(styles.root, className)}
          style={{ ...style, backgroundImage: `url(${src})` }}
        />
      );
    }
  )
);
Avatar.displayName = 'Avatar';
