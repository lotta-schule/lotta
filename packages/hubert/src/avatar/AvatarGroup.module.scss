.root {
  --deletable-button-offset: 4px;

  display: inline-flex;
  .avatar {
    margin-right: calc(var(--deletable-button-offset) * -1);
    &:last-of-type {
      margin-right: var(--deletable-button-offset);
    }
  }
  .overshoot {
    border-radius: 50%;
    background-color: rgb(var(--lotta-accent-grey-color));
    height: 30px;
    width: 30px;
    z-index: 1;
    font-family: var(--lotta-text-font-family);
    color: rgb(var(--lotta-text-color));
    box-shadow: 1px 1px 7px #0000002e;
    line-height: 2em;
    text-align: center;
  }
}
