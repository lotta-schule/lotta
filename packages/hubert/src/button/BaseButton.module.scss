.root {
  position: relative;
  border: 1px solid rgba(var(--lotta-primary-color), 1);
  color: rgba(var(--lotta-primary-color), 1);
  background: rgba(var(--lotta-box-background-color), 1);
  border-radius: 0.5em;
  padding: 0.7em;
  display: flex;
  text-transform: uppercase;
  letter-spacing: 1px;
  text-decoration: none;
  font-size: 0.8rem;
  justify-content: space-evenly;
  font-family: var(--lotta-text-font-family);
  &:not([disabled]) {
    cursor: pointer;
    &:hover,
    &:active {
      transition: linear;
      transition-duration: 0.2s;
      background-color: rgba(var(--lotta-primary-color), 0.1);
    }
  }

  &[disabled] {
    color: rgba(var(--lotta-disabled-color), 1);
    border-color: rgba(var(--lotta-disabled-color), 1);
  }

  &.selected {
    border: 1px solid rgba(var(--lotta-primary-color), 1);
    background-color: rgba(var(--lotta-primary-color), 0.2);
    filter: brightness(90%);

    &:hover {
      filter: brightness(90%);
    }
    &:active {
      filter: brightness(70%);
    }
  }

  &.fullWidth {
    width: 100%;
  }

  /**
      * Custom styles for different variants
      */
  &[data-variant='fill'] {
    background-color: rgba(var(--lotta-primary-color), 1);
    color: rgba(var(--lotta-contrast-text-color), 1);
    border: none;
    &:hover {
      transition: linear;
      transition-duration: 0.2s;
      background: rgba(var(--lotta-primary-color), 1);
      box-shadow: inset 0 0 20px #00000024;
    }
    &:active {
      background: rgba(var(--lotta-primary-color), 1);
      box-shadow: inset 0 0 20px #0000004f;
      filter: none;
    }
  }
  &[data-variant='borderless'] {
    border: none;
  }
  &[data-variant='error'] {
    background-color: rgba(var(--lotta-error-color), 1);
    color: #fff;
    border: none;
    &:hover {
      transition: linear;
      transition-duration: 0.2s;
      background-color: rgba(var(--lotta-error-color), 1);
      box-shadow: inset 0 0 20px #00000024;
    }
    &:active {
      background-color: rgba(var(--lotta-primary-color), 1);
      box-shadow: inset 0 0 20px #0000004f;
      filter: none;
    }
  }
}
