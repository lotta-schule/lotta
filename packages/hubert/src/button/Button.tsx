import * as React from 'react';
import { BaseButton, BaseButtonProps } from './BaseButton';
import { useButtonGroupContext } from './ButtonGroupContext';
import clsx from 'clsx';

import styles from './Button.module.scss';

export type ButtonProps = {
  /**
   * Button contents
   */
  label?: string;

  /**
   * Icon to show on the button side
   */
  icon?: React.ReactElement;

  /**
   * Disable setting a minimal height
   */
  small?: boolean;

  /**
   * Force style for button with only icon and without label,
   * even if child is found
   */
  onlyIcon?: boolean;

  /**
   * Add custom claaes to the element tree
   **/
  classes?: {
    root?: string;
    onlyIcon?: string;
    icon?: string;
    label?: string;
  };
} & BaseButtonProps;

export const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
  (
    { icon, label, onlyIcon, small, children, classes, className, ...props },
    ref
  ) => {
    const showOnlyIcon = onlyIcon || (icon && !(label || children));
    const { grouped } = useButtonGroupContext();

    return (
      <BaseButton
        {...props}
        ref={ref}
        className={clsx(
          styles.root,
          className,
          classes?.root,
          showOnlyIcon && [styles.onlyIcon, classes?.onlyIcon],
          {
            [styles.small]: small,
            [styles.grouped]: grouped,
          }
        )}
      >
        {icon && (
          <span className={clsx(styles.icon, classes?.icon)}>{icon}</span>
        )}
        {(label ?? children) && (
          <span className={clsx(styles.label, classes?.label)}>
            {label ?? children}
          </span>
        )}
      </BaseButton>
    );
  }
);
Button.displayName = 'Button';
