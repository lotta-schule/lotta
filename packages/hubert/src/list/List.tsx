import * as React from 'react';
import clsx from 'clsx';

import styles from './List.module.scss';

export type ListProps = Omit<React.HTMLProps<HTMLUListElement>, 'ref'>;

export const List = React.forwardRef(
  (
    { children, className, ...props }: ListProps,
    ref: React.Ref<HTMLUListElement>
  ) => {
    return (
      <ul className={clsx(styles.root, className)} {...props} ref={ref}>
        {children}
      </ul>
    );
  }
);
List.displayName = 'List';
