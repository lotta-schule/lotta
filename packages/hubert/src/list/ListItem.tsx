import * as React from 'react';
import { Divider } from '../divider/Divider';
import clsx from 'clsx';

import styles from './List.module.scss';

export type ListItemProps = React.HTMLProps<HTMLLIElement> & {
  leftSection?: React.ReactNode;
  rightSection?: React.ReactNode;
  isDivider?: boolean;
  isDisabled?: boolean;
  isFocused?: boolean;
  isSelected?: boolean;
};

export const ListItem = React.forwardRef(
  (
    {
      children,
      className,
      isDisabled,
      isFocused,
      isSelected,
      leftSection,
      rightSection,
      ...props
    }: ListItemProps,
    ref: React.ForwardedRef<HTMLLIElement>
  ) => {
    if (props.isDivider === true) {
      return (
        <li
          className={clsx(styles.li, styles.isDivider, className)}
          {...Object.fromEntries(
            Object.entries(props).filter(([key]) => key !== 'isDivider')
          )}
        >
          <Divider />
        </li>
      );
    }

    return (
      <li
        className={clsx(styles.li, className, {
          [styles.isDisabled]: isDisabled,
          [styles.isFocused]: isFocused,
          [styles.isSelected]: isSelected,
        })}
        {...(isFocused ? { ['data-is-focused']: true } : {})}
        ref={ref}
        {...props}
      >
        {leftSection && <div>{leftSection}</div>}
        <div className={styles.mainSection}>{children}</div>
        {rightSection && <div>{rightSection}</div>}
      </li>
    );
  }
);
ListItem.displayName = 'ListItem';
