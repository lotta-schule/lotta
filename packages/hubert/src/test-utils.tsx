import * as React from 'react';
import { DefaultThemes } from '@lotta-schule/theme';
import { render, RenderOptions } from '@testing-library/react';
import { MotionConfig } from 'framer-motion';
import { HubertProvider } from './HubertProvider';
import { GlobalStyles } from './theme';

const theme = DefaultThemes.standard;

const Wrapper = ({ children }: any) => (
  <HubertProvider>
    <MotionConfig transition={{ duration: 0 }}>
      <GlobalStyles theme={theme} />
      {children}
    </MotionConfig>
  </HubertProvider>
);

const customRender = (
  ui: React.ReactElement,
  renderOptions: Omit<RenderOptions, 'wrapper'> = {}
) =>
  render(ui, {
    wrapper: Wrapper,
    ...renderOptions,
  });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
