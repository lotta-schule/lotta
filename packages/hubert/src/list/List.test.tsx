import * as React from 'react';
import { render } from '../test-utils';
import { Avatar } from '../avatar/Avatar';
import { Button } from '../button/Button';
import { List } from './List';
import { ListItem } from './ListItem';

describe('List', () => {
  it('should render', () => {
    const screen = render(
      <List>
        <ListItem>Test</ListItem>
        <ListItem>Test</ListItem>
        <ListItem>Test</ListItem>
        <ListItem>Test</ListItem>
      </List>
    );
    expect(screen.getByRole('list')).toMatchSnapshot();
  });

  it('should render with subsection left', () => {
    const screen = render(
      <List>
        <ListItem leftSection={<Avatar src={''} />}>Test</ListItem>
        <ListItem leftSection={<Avatar src={''} />}>Test</ListItem>
        <ListItem leftSection={<Avatar src={''} />}>Test</ListItem>
        <ListItem leftSection={<Avatar src={''} />}>Test</ListItem>
      </List>
    );
    expect(screen.getByRole('list')).toMatchSnapshot();
  });

  it('should render with subsection right', () => {
    const screen = render(
      <List>
        <ListItem
          leftSection={<Avatar src={''} />}
          rightSection={<Button>Test</Button>}
        >
          Test
        </ListItem>
        <ListItem rightSection={<Button>Test</Button>}>Test</ListItem>
        <ListItem rightSection={<Button>Test</Button>}>Test</ListItem>
        <ListItem rightSection={<Button>Test</Button>}>Test</ListItem>
      </List>
    );
    expect(screen.getByRole('list')).toMatchSnapshot();
  });
});
