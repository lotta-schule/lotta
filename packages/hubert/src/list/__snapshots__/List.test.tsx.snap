// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List should render 1`] = `
<ul
  class="root"
>
  <li
    class="li"
  >
    <div
      class="mainSection"
    >
      Test
    </div>
  </li>
  <li
    class="li"
  >
    <div
      class="mainSection"
    >
      Test
    </div>
  </li>
  <li
    class="li"
  >
    <div
      class="mainSection"
    >
      Test
    </div>
  </li>
  <li
    class="li"
  >
    <div
      class="mainSection"
    >
      Test
    </div>
  </li>
</ul>
`;

exports[`List should render with subsection left 1`] = `
<ul
  class="root"
>
  <li
    class="li"
  >
    <div>
      <div
        class="root"
        role="img"
        style="background-image: url();"
      />
    </div>
    <div
      class="mainSection"
    >
      Test
    </div>
  </li>
  <li
    class="li"
  >
    <div>
      <div
        class="root"
        role="img"
        style="background-image: url();"
      />
    </div>
    <div
      class="mainSection"
    >
      Test
    </div>
  </li>
  <li
    class="li"
  >
    <div>
      <div
        class="root"
        role="img"
        style="background-image: url();"
      />
    </div>
    <div
      class="mainSection"
    >
      Test
    </div>
  </li>
  <li
    class="li"
  >
    <div>
      <div
        class="root"
        role="img"
        style="background-image: url();"
      />
    </div>
    <div
      class="mainSection"
    >
      Test
    </div>
  </li>
</ul>
`;

exports[`List should render with subsection right 1`] = `
<ul
  class="root"
>
  <li
    class="li"
  >
    <div>
      <div
        class="root"
        role="img"
        style="background-image: url();"
      />
    </div>
    <div
      class="mainSection"
    >
      Test
    </div>
    <div>
      <button
        aria-current="false"
        class="root root"
        data-variant="default"
        role="button"
        type="button"
      >
        <span
          class="label"
        >
          Test
        </span>
      </button>
    </div>
  </li>
  <li
    class="li"
  >
    <div
      class="mainSection"
    >
      Test
    </div>
    <div>
      <button
        aria-current="false"
        class="root root"
        data-variant="default"
        role="button"
        type="button"
      >
        <span
          class="label"
        >
          Test
        </span>
      </button>
    </div>
  </li>
  <li
    class="li"
  >
    <div
      class="mainSection"
    >
      Test
    </div>
    <div>
      <button
        aria-current="false"
        class="root root"
        data-variant="default"
        role="button"
        type="button"
      >
        <span
          class="label"
        >
          Test
        </span>
      </button>
    </div>
  </li>
  <li
    class="li"
  >
    <div
      class="mainSection"
    >
      Test
    </div>
    <div>
      <button
        aria-current="false"
        class="root root"
        data-variant="default"
        role="button"
        type="button"
      >
        <span
          class="label"
        >
          Test
        </span>
      </button>
    </div>
  </li>
</ul>
`;
